subprojects/launcher/src/main/java/org/gradle/tooling/internal/provider/ContinuousBuildActionExecutor.java

```diff
@@ -139,37 +140,47 @@ private void waitForDeployments(BuildAction action, BuildRequestMetaData request
     private BuildActionRunner.Result executeMultipleBuilds(BuildAction action, BuildRequestMetaData requestContext, BuildSessionContext buildSession,
                                                            BuildCancellationToken cancellationToken, CancellableOperationManager cancellableOperationManager, ContinuousExecutionGate continuousExecutionGate) {
         BuildActionRunner.Result lastResult;
-        while (true) {
-            PendingChangesListener pendingChangesListener = listenerManager.getBroadcaster(PendingChangesListener.class);
-            final FileSystemChangeWaiter waiter = changeWaiterFactory.createChangeWaiter(new SingleFirePendingChangesListener(pendingChangesListener), cancellationToken, continuousExecutionGate);
-            try {
-                lastResult = executeBuildAndAccumulateInputs(action, waiter, buildSession);
+        PendingChangesListener pendingChangesListener = listenerManager.getBroadcaster(PendingChangesListener.class);
+        FileChangeListener changeBroadcaster = listenerManager.getBroadcaster(FileChangeListener.class);
+        virtualFileSystem.registerChangeBroadcaster(changeBroadcaster);
+        try {
+            while (true) {
+                ContinuousBuildTriggerHandler continuousBuildTriggerHandler = new ContinuousBuildTriggerHandler(
+                    buildInputHierarchyFactory.create(),
+                    new SingleFirePendingChangesListener(pendingChangesListener),
+                    cancellationToken,
+                    continuousExecutionGate
+                );
+                try {
+                    listenerManager.addListener(continuousBuildTriggerHandler);
+                    lastResult = executeBuildAndAccumulateInputs(action, continuousBuildTriggerHandler, buildSession);
+
+                    if (!continuousBuildTriggerHandler.hasAnyInputs()) {
+                        logger.println().withStyle(StyledTextOutput.Style.Failure).println("Exiting continuous build as no executed tasks declared file system inputs.");
+                        return lastResult;
+                    } else {
+                        cancellableOperationManager.monitorInput(operationToken -> {
+                            continuousBuildTriggerHandler.wait(
+                                () -> logger.println().println("Waiting for changes to input files of tasks..." + determineExitHint(requestContext))
+                            );
+                            if (!operationToken.isCancellationRequested()) {
+                                continuousBuildTriggerHandler.reportChanges(logger);
+                            }
+                        });
+                    }
+                } finally {
+                    listenerManager.removeListener(continuousBuildTriggerHandler);
+                }
```

Comment:
This business with the listeners (`PendingChangesListener`, `FileChangeListener` and `ContinuousBuildTriggerHandler` which itself is a `FileChangeListener` _and_ a `TaskInputListener`) is really hard to follow. I wonder if this complexity is necessary. We don't have to clean it up now, but would be nice to have a plan to do it in a followup PR perhaps.

Reply:
I separated the different listeners. I think it is clearer now.